src/tactic/norm_fin.lean

```diff
@@ -0,0 +1,265 @@
+  simp [fin.eq_iff_veq, bit0, fin.add_def, ←h', ←h]
+end
+
+lemma fin.prove_mod_bit1 (n k l m : ℕ) (hk : k < n + 1) (hl : l < n + 1)
+  (h : l + l + 1 = m) (h' : m % (n + 1) = k) :
+  (bit1 ⟨l, hl⟩ : fin (n + 1)) = (⟨k, hk⟩ : fin (n + 1)) :=
+begin
+  cases n,
+  { simp },
+  simp [fin.eq_iff_veq, bit1, bit0, fin.add_def, ←h', ←h]
+end
+
+lemma fin.prove_lt (n a b : ℕ) (ha : a < n) (hb : b < n) (h : a < b) :
+  (⟨a, ha⟩ : fin n) < ⟨b, hb⟩ :=
+fin.lt_iff_coe_lt_coe.mpr h
+
+section
+local attribute [semireducible] reflected
+meta instance {n : ℕ} : has_reflect (fin (n + 1)) := fin.reflect
+end
+
+/-- Evaluates an expression as a finite number
+if that expression represents a `fin 1`, which always gives back `0`. -/
+private meta def expr.to_fin_one : expr → option (fin 1)
+| e := do k ← e.to_nat, pure (fin.of_nat 0)
+
+/-- Evaluates an expression as a finite number,
+if that expression represents a numeral of a `fin (n + 2)`,
+or `fin.succ` of a numeral.cast_succ` of a numeral. -/
+private meta def expr.to_fin_succ {n : ℕ} : expr → option (fin (n + 2))
+| `(@fin.succ %%t %%e) := do k ← e.to_nat, m ← t.to_nat, pure (fin.succ (@fin.of_nat n (k % m)))
+| e := do k ← e.to_nat, pure (fin.of_nat k)
+
+/-- Evaluates an expression as a finite number,
+if that expression represents a numeral of a `fin (n + 1)`,
+or `fin.succ` of a numeral.cast_succ` of a numeral. -/
+protected meta def expr.to_fin : Π (n : ℕ), expr → option (fin (n + 1))
+| 0       := expr.to_fin_one
+| (n + 1) := expr.to_fin_succ
+
+/-- Evaluates an expression into a finite `fin (n + 1)` number, if that expression is built up from
+  numerals, +, *, /, ⁻¹  -/
+protected meta def expr.eval_fin (n : ℕ) : expr → option (fin (n + 1))
+| `(has_zero.zero) := some 0
+| `(has_one.one) := some 1
+| `(bit0 %%q) := (*) 2 <$> q.eval_fin
+| `(bit1 %%q) := (+) 1 <$> (*) 2 <$> q.eval_fin
+| `(%%a + %%b) := (+) <$> a.eval_fin <*> b.eval_fin
+| `(%%a * %%b) := (*) <$> a.eval_fin <*> b.eval_fin
+| _ := none
+
+/-- `expr.of_fin α q` embeds `q` as a numeral expression inside the type `α`.
+Lean will try to infer the correct type classes on `α`, and the tactic will fail if it cannot.
+This function is similar to `fin.mk_numeral` but it takes fewer hypotheses and is tactic valued.
+-/
+protected meta def expr.of_fin {n : ℕ} (α : expr) : fin (n + 1) → tactic expr
+| ⟨k, _⟩ := expr.of_nat α k
+
+namespace tactic
+namespace instance_cache
+
+/-- `c.of_nat q` embeds `q` as a numeral expression inside the type `α`.
+Lean will try to infer the correct type classes on `c.α`, and the tactic will fail if it cannot.
+This function is similar to `fin.mk_numeral` but it takes fewer hypotheses and is tactic valued.
+-/
+protected meta def of_fin {n : ℕ} (c : instance_cache) :
+  fin (n + 1) → tactic (instance_cache × expr)
+| ⟨k, hk⟩ := c.of_nat k
+
+open norm_num
+
+/-- Given `a`,`b` `fin n` numerals, proves `⊢ a < b`. -/
+meta def prove_lt_fin (ic : instance_cache) : expr → expr → tactic (instance_cache × expr)
```

Comment:
Why the `instance_cache` argument? It's not really used in this function, is it needed to fit some API? If anything it seems like this should take a `nat` instance cache.

Reply:
In fact, this function is only used in `prove_ne_fin` which is used nowhere. Is this dead code?

Reply:
`prove_ne_fin` will be used in #5637

Reply:
I modeled it on the `prove_lt_nat` definition in norm_num. But I guess since I am not using `ic.mk_app`, I can try to remove the `instance_cache` argument.